src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/ReadOperation.java

```diff
@@ -404,6 +409,55 @@ private void getCount(AsyncResultSet rs, Throwable error, AtomicLong counter) {
     }
   }
 
+  /**
+   * Run estimated count query and parse the result set
+   *
+   * @param queryExecutor
+   * @param simpleStatement
+   * @return
+   */
+  default Uni<CountResponse> estimateDocumentCount(
+      QueryExecutor queryExecutor, SimpleStatement simpleStatement) {
+    AtomicLong counter = new AtomicLong();
+    final CompletionStage<AsyncResultSet> async =
+        queryExecutor
+            .executeEstimatedCount(simpleStatement)
+            .whenComplete(
+                (rs, error) -> {
+                  getEstimatedCount(rs, error, counter);
+                });
+
+    return Uni.createFrom()
+        .completionStage(async)
+        .onItem()
+        .transform(
+            rs -> {
+              return new CountResponse(counter.get());
+            });
+  }
+
+  private void getEstimatedCount(AsyncResultSet rs, Throwable error, AtomicLong counter) {
+    if (error != null) {
+      throw new JsonApiException(ErrorCode.COUNT_READ_FAILED);
+    } else {
+
+      for (Row row : rs.currentPage()) {
+        Long rangeStart = Long.valueOf(row.getString("range_start"));
+        Long rangeEnd = Long.valueOf(row.getString("range_end"));
+        double rangeSize = 0;
+        if (rangeStart > rangeEnd) {
+          rangeSize = (MAX_TOKEN - rangeStart) + (rangeEnd - MIN_TOKEN);
```

Comment:
this is where I wonder if `long` vs `double` might not work well: they have same number of bits but accuracy of `double` actually only covers subset of integer positions I think (causing issues in Javascript where 64-bit `double` is the only numeric type, and cannot accurately express all of 64-bit `long` range)